cura/Machines/MaterialManager.py

```diff
@@ -0,0 +1,307 @@
+                self._diameter_machine_variant_material_map[approximate_diameter] = {}
+
+            machine_variant_material_map = self._diameter_machine_variant_material_map[approximate_diameter]
+            if definition not in machine_variant_material_map:
+                machine_variant_material_map[definition] = MaterialNode()
+
+            machine_node = machine_variant_material_map[definition]
+            variant_name = material_metadata.get("variant_name")
+            if not variant_name:
+                # if there is no variant, this material is for the machine, so put its metadata in the machine node.
+                machine_node.material_map[root_material_id] = MaterialNode(material_metadata)
+            else:
+                # this material is variant-specific, so we save it in a variant-specific node under the
+                # machine-specific node
+                if variant_name not in machine_node.children_map:
+                    machine_node.children_map[variant_name] = MaterialNode()
+
+                variant_node = machine_node.children_map[variant_name]
+                if root_material_id not in variant_node.material_map:
+                    variant_node.material_map[root_material_id] = MaterialNode(material_metadata)
+                else:
+                    # Sanity check: make sure we don't have duplicated variant-specific materials for the same machine
+                    raise RuntimeError("Found duplicate variant name [%s] for machine [%s] in material [%s]" %
+                                       (variant_name, definition, material_metadata["id"]))
+
+        self.materialsUpdated.emit()
+
+    def _updateMaps(self):
+        self.initialize()
+
+    def _onContainerMetadataChanged(self, container):
+        self._onContainerChanged(container)
+
+    def _onContainerChanged(self, container):
+        container_type = container.getMetaDataEntry("type")
+        if container_type != "material":
+            return
+
+        # update the maps
+        self._update_timer.start()
+
+    def getMaterialGroup(self, root_material_id: str) -> Optional[MaterialGroup]:
+        return self._material_group_map.get(root_material_id)
+
+    def getRootMaterialIDForDiameter(self, root_material_id: str, approximate_diameter: str) -> str:
+        return self._material_diameter_map.get(root_material_id).get(approximate_diameter, root_material_id)
+
+    def getRootMaterialIDWithoutDiameter(self, root_material_id: str) -> str:
+        return self._diameter_material_map.get(root_material_id)
+
+    def getMaterialGroupListByGUID(self, guid: str) -> Optional[list]:
+        return self._guid_material_groups_map.get(guid)
+
+    #
+    # Return a dict with all root material IDs (k) and ContainerNodes (v) that's suitable for the given setup.
+    #
+    def getAvailableMaterials(self, machine_definition_id: str, variant_name: Optional[str], diameter: float) -> dict:
```

Comment:
This is only the extruder variant name because we don't use the global variant name in material/quality lookup yet. I will rename this to `extruder_variant_name` to make this clear.

Reply:
Renamed in ea2b076db

Reply:
We do use the global variant for material look-up though, right?

For those who have access, this is an example: https://github.com/Ultimaker/fdm_materials_private/blob/46d99b18fb1cea920add5802937c499820bfa70c/generic_pla.xml.fdm_material#L94

Reply:
Okay, so it turns out that a single material profile contains the compatible/recommended metadata for all global variants, rather than creating a new profile for each global variant like we do for the other dimensions, and that was not part of this refactor. That's why the global variant parameter is not part of this call.